File: ios/Packages/GameCore/Sources/GameCore/Fog/FogLayer.swift
/// Битовая карта тумана одного тайла: 256 × 256 клеток, бит 1 — клетка открыта.
/// Такие же 8 КБ (до сжатия) хранит сервер в `fog_tiles.bits` (PLAN.md, §7.3).
public struct FogTileBits: Hashable, Codable, Sendable {
    public static let wordCount = 256 * 256 / 64

    public private(set) var words: [UInt64]

    public init() {
        words = Array(repeating: 0, count: Self.wordCount)
    }

    public init?(words: [UInt64]) {
        guard words.count == Self.wordCount else { return nil }
        self.words = words
    }

    public func isSet(_ bit: Int) -> Bool { words[bit >> 6] & (1 << UInt64(bit & 63)) != 0 }

    public mutating func set(_ bit: Int) { words[bit >> 6] |= 1 << UInt64(bit & 63) }

    /// Сколько клеток открыто.
    public var count: Int { words.reduce(0) { $0 + $1.nonzeroBitCount } }

    /// Сколько клеток открыто здесь и не было открыто в `old`: popcount(new & ~old).
    public func newCount(comparedTo old: FogTileBits) -> Int {
        zip(words, old.words).reduce(0) { $0 + ($1.0 & ~$1.1).nonzeroBitCount }
    }

    public mutating func formUnion(_ other: FogTileBits) {
        for index in words.indices {
            words[index] |= other.words[index]
        }
    }
}

/// Слой «Исследования» одного игрока: открытые клетки по тайлам (PLAN.md, §3.10).
///
/// Валидное движение открывает круг радиусом 25 м вокруг каждой точки пути; между точками путь
/// «закрашивается» сплошной полосой, если разрыв не больше порога (100 м пешком, 200 м на велосипеде) —
/// длинный разрыв значит, что GPS пропадал, и закрашивать неизвестный путь нельзя.
public struct FogLayer: Hashable, Codable, Sendable {
    public private(set) var tiles: [FogTileKey: FogTileBits] = [:]

    public init() {}

    public var isEmpty: Bool { tiles.isEmpty }

    /// Открывает круг радиусом `radius` метров вокруг точки.
    public mutating func reveal(around coordinate: Coordinate, radius: Double = 25) {
        let size = FogGrid.cellSizeMeters(atLatitude: coordinate.latitude)
        let (px, py) = FogGrid.pixel(of: coordinate)
        let reach = Int((radius / size).rounded(.up)) + 1
        let cx = Int(px.rounded(.down))
        let cy = Int(py.rounded(.down))
        for dy in -reach...reach {
            for dx in -reach...reach {
                let x = cx + dx
                let y = cy + dy
                // Клетка открыта, если её центр в круге.
                let ex = (Double(x) + 0.5 - px) * size
                let ey = (Double(y) + 0.5 - py) * size
                if ex * ex + ey * ey <= radius * radius {
                    set(FogCell(x: x, y: y))
                }
            }
        }
    }

    /// Открывает полосу вдоль пути между двумя точками (или только концы, если разрыв длиннее `maxGap`).
    public mutating func reveal(from a: Coordinate, to b: Coordinate, radius: Double = 25, maxGap: Double = 100) {
        let length = Geodesy.distance(from: a, to: b)
        guard length <= maxGap else {
            reveal(around: a, radius: radius)
            reveal(around: b, radius: radius)
            return
        }

        // Шаг — половина клетки, чтобы круги сливались в сплошную полосу без «зубцов».
        let step = FogGrid.cellSizeMeters(atLatitude: a.latitude) / 2
        let count = max(1, Int((length / step).rounded(.up)))
        for i in 0...count {
            let t = Double(i) / Double(count)
            let point = Coordinate(
                latitude: a.latitude + (b.latitude - a.latitude) * t,
                longitude: a.longitude + (b.longitude - a.longitude) * t
            )
            reveal(around: point, radius: radius)
        }
    }

    public func isRevealed(_ cell: FogCell) -> Bool { tiles[cell.tile]?.isSet(cell.bitIndex) ?? false }

    /// Всего открытых клеток.
    public var cellCount: Int { tiles.values.reduce(0) { $0 + $1.count } }

    /// Открытая площадь, м²: клетки каждого тайла × площадь клетки на широте его центра.
    public var areaSquareMeters: Double {
        tiles.reduce(0) { total, entry in
            let latitude = FogGrid.center(of: FogCell(x: (entry.key.x << 8) + 128, y: (entry.key.y << 8) + 128))
                .latitude
            let size = FogGrid.cellSizeMeters(atLatitude: latitude)
            return total + Double(entry.value.count) * size * size
        }
    }

    /// Сколько клеток открыто здесь и не было открыто в `old` («+N га сегодня»).
    public func newCellCount(comparedTo old: FogLayer) -> Int {
        tiles.reduce(0) { total, entry in
            total + entry.value.newCount(comparedTo: old.tiles[entry.key] ?? FogTileBits())
        }
    }

    public mutating func formUnion(_ other: FogLayer) {
        for (key, bits) in other.tiles {
            tiles[key, default: FogTileBits()].formUnion(bits)
        }
    }

    private mutating func set(_ cell: FogCell) {
        tiles[cell.tile, default: FogTileBits()].set(cell.bitIndex)
    }
}

File: backend/src/Gorodki.Domain/Fog/FogLayer.cs
using System.Buffers.Binary;
using System.Numerics;
using Gorodki.Domain.Geo;

namespace Gorodki.Domain.Fog;

/// <summary>
/// Битовая карта тумана одного тайла: 256 × 256 клеток, бит 1 — клетка открыта. На сервере хранится в <c>fog_tiles.bits</c>
/// (PLAN.md, §7.3): 8 КБ до сжатия.
/// </summary>
public sealed class FogTileBits
{
    public const int WordCount = 256 * 256 / 64;

    public const int ByteCount = WordCount * 8;

    private readonly ulong[] _words = new ulong[WordCount];

    public IReadOnlyList<ulong> Words => _words;

    public bool IsSet(int bit) => (_words[bit >> 6] & (1UL << (bit & 63))) != 0;

    public void Set(int bit) => _words[bit >> 6] |= 1UL << (bit & 63);

    /// <summary>Сколько клеток открыто.</summary>
    public int Count => _words.Sum(BitOperations.PopCount);

    /// <summary>Сколько клеток открыто здесь и не было в <paramref name="old"/>: popcount(new &amp; ~old).</summary>
    public int NewCount(FogTileBits old)
    {
        var total = 0;
        for (var i = 0; i < WordCount; i++)
        {
            total += BitOperations.PopCount(_words[i] & ~old._words[i]);
        }

        return total;
    }

    /// <summary>Сколько клеток открыто и здесь, и в <paramref name="other"/>: popcount(this &amp; other).</summary>
    public int CountAnd(FogTileBits other)
    {
        var total = 0;
        for (var i = 0; i < WordCount; i++)
        {
            total += BitOperations.PopCount(_words[i] & other._words[i]);
        }

        return total;
    }

    public void UnionWith(FogTileBits other)
    {
        for (var i = 0; i < WordCount; i++)
        {
            _words[i] |= other._words[i];
        }
    }

    /// <summary>8 192 байта, слова little-endian — так тайл хранится (до сжатия).</summary>
    public byte[] ToBytes()
    {
        var bytes = new byte[ByteCount];
        for (var i = 0; i < WordCount; i++)
        {
            BinaryPrimitives.WriteUInt64LittleEndian(bytes.AsSpan(i * 8), _words[i]);
        }

        return bytes;
    }

    public static FogTileBits FromBytes(ReadOnlySpan<byte> bytes)
    {
        if (bytes.Length != ByteCount)
        {
            throw new FormatException($"Тайл тумана — ровно {ByteCount} байт.");
        }

        var bits = new FogTileBits();
        for (var i = 0; i < WordCount; i++)
        {
            bits._words[i] = BinaryPrimitives.ReadUInt64LittleEndian(bytes[(i * 8)..]);
        }

        return bits;
    }
}

/// <summary>
/// Слой «Исследования» одного игрока (PLAN.md, §3.10) — перенос <c>FogLayer</c> из GameCore строка в строку. Валидное движение
/// открывает круг вокруг каждой точки пути, между точками — сплошную полосу, если разрыв не больше порога лиги.
/// </summary>
public sealed class FogLayer
{
    private readonly SortedDictionary<FogTileKey, FogTileBits> _tiles = new();

    public IReadOnlyDictionary<FogTileKey, FogTileBits> Tiles => _tiles;

    public int CellCount => _tiles.Values.Sum(t => t.Count);

    /// <summary>Открывает круг радиусом <paramref name="radius"/> метров вокруг точки: клетка открыта, если её центр в круге.</summary>
    public void RevealAround(double latitude, double longitude, double radius)
    {
        var size = FogGrid.CellSizeMeters(latitude);
        var (px, py) = FogGrid.Pixel(latitude, longitude);
        var reach = (int)Math.Ceiling(radius / size) + 1;
        var cx = (int)Math.Floor(px);
        var cy = (int)Math.Floor(py);
        for (var dy = -reach; dy <= reach; dy++)
        {
            for (var dx = -reach; dx <= reach; dx++)
            {
                var x = cx + dx;
                var y = cy + dy;
                var ex = (x + 0.5 - px) * size;
                var ey = (y + 0.5 - py) * size;
                if ((ex * ex) + (ey * ey) <= radius * radius)
                {
                    Set(new FogCell(x, y));
                }
            }
        }
    }

    /// <summary>Открывает полосу вдоль пути между двумя точками (или только концы, если разрыв длиннее <paramref name="maxGap"/>).</summary>
    public void RevealPath(double latitudeA, double longitudeA, double latitudeB, double longitudeB, double radius, double maxGap)
    {
        var length = Geodesy.Distance(latitudeA, longitudeA, latitudeB, longitudeB);
        if (length > maxGap)
        {
            RevealAround(latitudeA, longitudeA, radius);
            RevealAround(latitudeB, longitudeB, radius);
            return;
        }

        // Шаг — половина клетки, чтобы круги сливались в сплошную полосу без «зубцов».
        var step = FogGrid.CellSizeMeters(latitudeA) / 2;
        var count = Math.Max(1, (int)Math.Ceiling(length / step));
        for (var i = 0; i <= count; i++)
        {
            var t = (double)i / count;
            RevealAround(latitudeA + ((latitudeB - latitudeA) * t), longitudeA + ((longitudeB - longitudeA) * t), radius);
        }
    }

    public bool IsRevealed(FogCell cell) => _tiles.TryGetValue(cell.Tile, out var bits) && bits.IsSet(cell.BitIndex);

    public void UnionWith(FogLayer other)
    {
        foreach (var (key, bits) in other._tiles)
        {
            TileAt(key).UnionWith(bits);
        }
    }

    /// <summary>Тайл для записи (создаётся пустым) — например, чтобы загрузить сохранённые биты.</summary>
    public FogTileBits TileAt(FogTileKey key)
    {
        if (!_tiles.TryGetValue(key, out var bits))
        {
            bits = new FogTileBits();
            _tiles[key] = bits;
        }

        return bits;
    }

    private void Set(FogCell cell) => TileAt(cell.Tile).Set(cell.BitIndex);
}
